### PR TITLE
feat: wait for transaction confirmations

### PR DESCRIPTION
**Pull request by unparalleled-js**
### What I did

fixes: #274 

### How I did it

Add parameter to Transaction for setting the required confirmations
Add default required confirmations to the `NetworkAPI`
If transaction was not given a `required_confirmations` kwarg, use the default from the `NetworkAPI`

### How to verify it

* Run transactions on development - should be no confirmations needed
* Run transactions on testnets - should wait for 1 transaction
* Try adjusting the required confirmations on transfers, deploys, and contract txns

### Checklist

- [ ] Passes all linting checks (pre-commit and CI jobs)
- [ ] New test cases have been added and are passing
- [ ] Documentation has been updated
- [ ] PR title follows [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) standard (will be automatically included in the changelog)

### CODE REVIEW

File: src/ape/api/providers.py [hunk 1]
```diff
@@ -299,14 +301,64 @@ def send_call(self, txn: TransactionAPI) -> bytes:
...
+        receipt = self.network.ecosystem.receipt_class.decode({**txn, **receipt_data})
+
+        # Wait for nonce to catch up in case provider is being slow.
+        sender_nonce = self.get_nonce(txn["from"])
+        while sender_nonce <= txn.nonce:  # type: ignore
+            logger.debug(f"Waiting for provider nonce to catch up (provider={self.name})")
+            sender_nonce = self.get_nonce(txn["from"])
+            time.sleep(1)
+
+        if not required_confirmations:
```
Comment (fubuloubu) [inline src/ape/api/providers.py:323]:
```suggestion
        if required_confirmations == 0:
```

File: src/ape/api/providers.py [hunk 2]
```diff
@@ -30,6 +31,7 @@ class TransactionAPI:
     gas_limit: Optional[int] = None  # NOTE: `Optional` only to denote using default behavior
     data: bytes = b""
     type: TransactionType = TransactionType.STATIC
+    required_confirmations: Optional[int] = None
```
Comment (unparalleled-js) [inline src/ape/api/providers.py:34]: - [ ] Mention how this can be used to override the network confirmations

File: src/ape/api/providers.py [hunk 3]
```diff
@@ -299,14 +301,64 @@ def send_call(self, txn: TransactionAPI) -> bytes:
...
+            elif confirmations_remaining == 1:
+                message = f"'{confirmations_remaining}' confirmation remains."
+            else:
+                message = f"'{confirmations_remaining}' confirmations remaining."
+
+            logger.info(message)
+            logged_confirmations.append(confirmation)
+
+        log_confirmation(0)
+        last_logged_confirmations = 0
```
Comment (fubuloubu) [inline src/ape/api/providers.py:344]: use a progressbar?

**Comment by unparalleled-js** [comment]
everything in https://github.com/ApeWorX/ape/pull/279 is also in this one

File: src/ape/api/providers.py [hunk 4]
```diff
@@ -102,6 +103,32 @@ class TransactionStatusEnum(IntEnum):
     NO_ERROR = 1
 
 
+class ConfirmationsProgressBar:
+    """
+    A progress bar tracking the confirmations of a
+    transacitons.
```
Comment (fubuloubu) [inline src/ape/api/providers.py:109]:
```suggestion
    A progress bar tracking the confirmations of a transaction
```

**Comment by unparalleled-js** [comment]
- [ ] A todo that is left is the config file way of editing the defaults...

**Review by fubuloubu — DISMISSED** [review]

**Comment by fubuloubu** [comment]
approved this one by accident lol

**Comment by unparalleled-js** [comment]
I wish there was an easy way to get `wait_for_transaction_receipt()` to be part of the "Receipt Promise" type logic, so we can Ctrl+C out of that part too.. Because that can take a while sometimes.

**Comment by unparalleled-js** [comment]
The ctrl C out of waiting for confs does not work yet but everything else does I believe

**Review by fubuloubu — DISMISSED** [review]

**Review by evanblank3 — APPROVED** [review]
Thanks for the walkthrough!

**Review by evanblank3 — APPROVED** [review]

**Pull request merged**